xarray/core/variable.py

```diff
@@ -1212,35 +1212,94 @@ def __setitem__(self, key, value):
         raise TypeError('%s values cannot be modified' % type(self).__name__)
 
     @classmethod
-    def concat(cls, variables, dim='concat_dim', positions=None,
-               shortcut=False):
-        """Specialized version of Variable.concat for IndexVariable objects.
-
-        This exists because we want to avoid converting Index objects to NumPy
-        arrays, if possible.
+    def concat_numpy(cls, variables, positions=None):
         """
-        if not isinstance(dim, basestring):
-            dim, = dim.dims
+        Concatenates variables. Works for variables whose dtype is
+        different from numpy.object. If variables' dtype is numpy.object
+        it throws TypeError and "concat" function will use
+        concat_pandas function
+        :param variables: list of variables to concatenate
+        :return: Concatenated variables
+        """
+        variable_type_set = set(map(lambda v: type(v.data), variables))
 
-        variables = list(variables)
-        first_var = variables[0]
+        if len(variable_type_set) > 1:
+            raise TypeError('Trying to concatenate variables of '
```

Comment:
These extra checks would break some things that currently works with `np.concatenate` (e.g., `concatenate([float, int]) -> float`). So I would remove them, and replace this with the simpler `np.concatenate([v.data for v in variables])`.